// next/dist/esm/shared/lib/lazy-dynamic/loadable.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Suspense, lazy, Fragment } from "react";
import { NoSSR } from "./dynamic-no-ssr";
// Normalize loader to return the module as form { default: Component } for `React.lazy`.
// Also for backward compatible since next/dynamic allows to resolve a component directly with loader
// Client component reference proxy need to be converted to a module.
function convertModule(mod) {
    return {
        default: (mod == null ? void 0 : mod.default) || mod
    };
}
function Loadable(options) {
    const opts = {
        loader: null,
        loading: null,
        ssr: true,
        ...options
    };
    const loader = ()=>opts.loader != null ? opts.loader().then(convertModule) : Promise.resolve(convertModule(()=>null));
    const Lazy = /*#__PURE__*/ lazy(loader);
    const Loading = opts.loading;
    const Wrap = opts.ssr ? Fragment : NoSSR;
    function LoadableComponent(props) {
        const fallbackElement = Loading ? /*#__PURE__*/ _jsx(Loading, {
            isLoading: true,
            pastDelay: true,
            error: null
        }) : null;
        return /*#__PURE__*/ _jsx(Suspense, {
            fallback: fallbackElement,
            children: /*#__PURE__*/ _jsx(Wrap, {
                children: /*#__PURE__*/ _jsx(Lazy, {
                    ...props
                })
            })
        });
    }
    LoadableComponent.displayName = "LoadableComponent";
    return LoadableComponent;
}
export default Loadable;
